src/main.rs

```diff
@@ -83,29 +114,72 @@ fn run() -> Result<()> {
         })
         .collect();
 
-    let mut port = match matching_ports.len() {
+    match matching_ports.len() {
         0 => {
             return Err(
                 "no matching USB serial device found.\n       Remember to put the \
                                  device in bootloader mode by pressing the reset button!"
-                    .to_string()
                     .into(),
             )
         }
-        1 => {
-            let port = &matching_ports[0].port_name;
-            log::debug!("opening {} (type {:?})", port, matching_ports[0].port_type);
-            serialport::new(port, 115200)
-                .timeout(Duration::from_millis(1000))
-                .open()?
-        }
+        1 => (),
         _ => {
-            return Err("multiple matching USB serial devices found"
-                .to_string()
-                .into())
+            if !args.all {
+                return Err("multiple matching USB serial devices found".into());
+            }
         }
     };
 
+    let mut errors_in_all = false;
+
+    for port in matching_ports {
+        let result = run_on_port(&port, &args, image.as_deref());
+        if let Err(e) = result {
+            if args.all {
+                // The current port is printed in run_on_port anyway, but it doesn't hurt to be
+                // explicit, especially since stdout and stderr might not be sorted properly.
+                eprintln!("error processing {}: {e}", &port.port_name);
+                errors_in_all = true;
+            } else {
+                return Err(e);
+            }
+        }
+    }
+
+    if errors_in_all {
+        return Err(PreviousErrors.into());
+    }
+
+    if image.is_none() && !args.get_images && !args.abort {
+        // This is done at the end so that errors from working on an --all still show up, to
+        // increase the usefulness of RUST_LOG=debug or as a kind of readiness check.
+        return Err("No actions performed; provide an .elf file on the command line to flash, or set querying options.".into());
+    }
+
+    Ok(())
+}
+
+fn run_on_port(port: &serialport::SerialPortInfo, args: &Args, image: Option<&[u8]>) -> Result<()> {
+    log::debug!("opening {} (type {:?})", port.port_name, port.port_type);
+    if args.all {
+        let serial = match &port.port_type {
+            serialport::SerialPortType::UsbPort(s) => s.serial_number.as_ref(),
+            // Those don't get through filtering anyway
+            _ => None,
+        };
+        println!(
+            "Port {} (serial: {})",
```

Comment:
Perhaps `"Found port {} (serial: {})"`, so it reads like it's looping through a list of things that it found?